Charge in arrears

rorepo opened the pull request:
Implemented the core part of the solution for issue "charge minimum in arrears, not in advance #3378" 
Overview of the solution:
i) Added a new column giving_due in payment_instructions. To be populated/incremented with the corresponding amount, each time the amount is not charged due to being under the minimum charge threshold
ii) After creation of payday_payment_instructions, the giving_due value for each participant-team combination is updated with the sum of giving_due values for that combination, from payment_instructions. 
iii) The 'giving_today' value for each participant is calculated as sum(amount + giving_due) from all records for that participant in payday_payment_instructions
iv) The 'giving_today' value is checked against the MINIMUM_CHARGE threshold before creating card holds if at/above the threshold. Have a doubt here - whether it's the basic amount or the upcharged amount that is to be used in the check
v) In the case of participants who have been charged, all giving_due values in payment_instructions for those participants are reset to 0
vi) In the case of participants who have not been charged, the giving_due value in the current record in payment_instructions is updated to amount + giving_due

Will await comments/suggestions before proceeding further. 

Further changes planned:
i) Add a new column giving_due in participants, to cache the total giving_due value for participants
ii) Update the UI to include the above

whit537 commented:
!m @rorepo

When you're ready, go ahead and apply the "Review" lable.

whit537 commented on gratipay/billing/payday.py:
```diff
@@ -266,6 +271,26 @@ def process_payment_instructions(cursor):
+        charge threshold, park this for the next week by adding the current subscription amount 
+        to giving_due
+        """
+	cursor.run("""INSERT INTO participants_payments_uncharged 
```
4-space indents instead of tabs, please.

whit537 commented on gratipay/billing/payday.py:
```diff
@@ -308,7 +333,7 @@ def settle_card_holds(self, cursor, holds):
               FROM payday_participants
              WHERE new_balance < 0
-        """)
+        """,dict(MINIMUM_CHARGE=MINIMUM_CHARGE))
```
Passing in the `dict` is extraneous, no?

whit537 replied:
Removed in e087c60987c523b66902a196d4ed96ee511c1b34.

whit537 commented:
Reformatting for readability. :-)

----

Implemented the core part of the solution for issue "charge minimum in arrears, not in advance #3378" 
Overview of the solution:

1. Added a new column `giving_due` in `payment_instructions`. To be populated/incremented with the corresponding amount, each time the amount is not charged due to being under the minimum charge threshold
1. After creation of `payday_payment_instructions`, the `giving_due` value for each participant-team combination is updated with the sum of `giving_due` values for that combination, from `payment_instructions`. 
1. The `giving_today` value for each participant is calculated as `sum(amount + giving_due)` from all records for that participant in `payday_payment_instructions`
1. The `giving_today` value is checked against the `MINIMUM_CHARGE` threshold before creating card holds if at/above the threshold. Have a doubt here - whether it's the basic amount or the upcharged amount that is to be used in the check
1. In the case of participants who have been charged, all `giving_due` values in `payment_instructions` for those participants are reset to 0
1. In the case of participants who have not been charged, the `giving_due` value in the current record in `payment_instructions` is updated to `amount + giving_due`

Will await comments/suggestions before proceeding further. 

Further changes planned:

1. Add a new column `giving_due` in participants, to cache the total `giving_due` value for participants
1. Update the UI to include the above

whit537 commented:
> Have a doubt here - whether it's the basic amount or the upcharged amount that is to be used in the check

I think it should be the basic amount, because that will be more natural to people. "I give 50¢/wk, so I'll get charged every 20 weeks," rather than getting charged in the 19th week because the fee puts you over the top. I mean, that's pretty minor. It's implemented as checking the basic amount right now, iirc.

whit537 commented:
Wtf?

![screen shot 2015-08-13 at 4 01 04 pm](https://cloud.githubusercontent.com/assets/134455/9260243/88881834-41d4-11e5-8f35-3cc17b86bb55.png)

whit537 commented:
We're going to need at least a few tests for this.

whit537 commented:
I mean for this PR as a whole. I'm not seeing any tests for the new functionality.

whit537 commented on sql/payday.sql:
```diff
@@ -76,10 +76,24 @@ CREATE INDEX ON payday_payment_instructions (participant);
+CREATE TABLE participants_payments_uncharged AS
+    SELECT participant 
+      FROM payday_payment_instructions
+     WHERE 1 = 2;
```
What's going on with this construction? One does not equal two. :-)

whit537 replied:
https://github.com/gratipay/gratipay.com/pull/3675#issuecomment-131705592:

> The '1 = 2' thing is an old trick I've used quite a lot - when I wished to copy the structure (full or part) of some table, without the data. Might be overkill here for just one column, but for one thing, all other table creations in this script are in the form of selects and also, this usage means it will not break if the participant column is modified in payday_payment_instructions/payment_instructions.

rohitpaulk commented:
<img width="455" alt="screen shot 2015-08-15 at 4 20 59 pm" src="https://cloud.githubusercontent.com/assets/3893573/9288551/ab7e25cc-4369-11e5-8de6-9e3787597749.png">

@whit537 - Can you try re-enabling the repo at Travis?

rorepo commented:
Have corrected the formatting and the extraneous argument. 

The '1 = 2' thing is an old trick I've used quite a lot - when I wished to copy the structure (full or part) of some table, without the data. Might be overkill here for just one column, but for one thing, all other table creations in this script are in the form of selects and also, this usage means it will not break if the participant column is modified in payday_payment_instructions/payment_instructions.

Working on the tests now. Haven't used mock or vcr before, so took a while to figure these out.

whit537 commented:
> Have corrected the formatting and the extraneous argument.

@rorepo Note that I pushed a couple commits towards this end as well: 22867fd32e8eca30b271fbc340df9fe75dd54372 & e087c60987c523b66902a196d4ed96ee511c1b34.

whit537 commented:
> Can you try re-enabling the repo at Travis?

It appears to be re-enabled already. O.O

The failure on e087c60987c523b66902a196d4ed96ee511c1b34 was spurious ("GitHub rate limit exceeded"), so I restarted that build.

whit537 commented:
> Working on the tests now. Haven't used mock or vcr before, so took a while to figure these out.

Cool. You almost certainly don't need vcr to test this code, and probably not even mock.

rorepo commented:
@whit537 - I'm afraid I didn't get your comment about not needing vcr/mock. The way I figured it, I'd have to update test_billing_payday.py, which uses these. I'd have to add new tests as well as update existing ones, because the 'charge-in-arrears' functionality would make some of these invalid e.g. 'test_payday_moves_money', which was for a payment of $6 and which would now not be charged since this is now below the charge threshold.

whit537 commented:
@rorepo I guess you're right. :-)

I was just going on a hunch, since you're actually touching code, don't mind me. Sorry! :-)

rorepo commented:
Looking at the tests exposed a few bugs as well as the fact that quite a few of the existing tests would have to change. 

The key points in tests:
* test_billing_payday.py: ```test_payday_moves_money``` needed to be replaced by two new methods - ```test_payday_moves_money_above_min_charge``` and ```test_payday_does_not_move_money_below_min_charge```, to test the behaviour when the amount to be charged is **>= MINIMUM_CHARGE** and **< MINIMUM_CHARGE**, respectively
* Other tests have needed related changes. Primarily, to increase amount to **>= MINIMUM_CHARGE**, to make sure the test works as intended

@whit537 I've removed the method ```park_payment_instructions``` from payday.py. This had actually been unused - I'd forgotten to remove it before pushing

whit537 commented:
The commit log on this PR looks weird. Why are there two of some commits? Why is the 1880 version bump on here?

whit537 commented:
I think maybe we need a rebase of this branch to clean up the commits.

whit537 commented:
I pulled out the first commit, 13faa2a435f2e5331a8147533eaebc79ebf9899a, into a new PR (#3695). I'm getting bit by #3579 over on #3694, and also had this fix on #3568. Let's take care of this, already. :-)

whit537 commented:
Rebasing on master got rid of the 1880 version bump. Now for the duplicate commits ...

rorepo commented:
Don't quite understand, @whit537. I did do a rebase before merging and I'd assume that's how the 1880 version came in.
Also, which are the duplicated commits you mean? If you mean the code formatting and cleanup, the latest one is what I did following some errors/warnings shown during the rebase about trailing spaces. Also removed an 'import time' I'd put in in between while running ```test_history.py```.

whit537 commented:
@rorepo After rebasing on master, here are the commits I'm seeing:

![screen shot 2015-08-19 at 1 59 02 pm](https://cloud.githubusercontent.com/assets/134455/9364461/7da9cb42-467a-11e5-9281-dc4e4ebd08c1.png)

It *looks* like we have duplicate commits for:

 - Preliminary commit - charging in arrears
 - Schema changes moved to branch.sql
 - Clean up code formatting
 - Remove extraneous argument

whit537 commented:
> Don't quite understand, @whit537.

Neither do I. :-) My git-fu is not that strong.

I think at this point we should rebase to squash the duplicate commits, and then make sure the overall diff is as expected, and then forge ahead. I'm going to rebase ... unless you object?

whit537 commented:
`$ git rebase -i f887fdd5f8d146d4c2ff5849143fe38059363c3e`

... presents me with ...

```
pick 97ead4e Preliminary commit - charging in arrears                                                         
pick 5b58b70 Schema changes moved to branch.sql                                                               
pick 14045c6 Clean up code formatting                                                                         
pick d9980f5 Remove extraneous argument                                                                       
pick 816776f Preliminary commit - charging in arrears                                                         
pick af42c27 Schema changes moved to branch.sql                                                               
pick 005573e Clean up code formatting                                                                         
pick fe2c70f Remove extraneous argument                                                                       
pick 1b3a373 Refined and added tests                                                                          
pick 98f0f3e Code formatting and cleanup                                                                      
                                                                                                              
# Rebase f887fdd..98f0f3e onto f887fdd
```

rorepo commented:
You're welcome to proceed, @whit537. My git-fu is almost certainly weaker than yours, in any case :-) 
Saw the duplicates. Hadn't checked that far back earlier.

whit537 commented:
> You're welcome to proceed, @whit537.

Okay, working on it ...

whit537 commented:
Here's what I'm trying to get to work:

```
pick 97ead4e Preliminary commit - charging in arrears                                                         
fixup 816776f Preliminary commit - charging in arrears                                                        
pick 5b58b70 Schema changes moved to branch.sql                                                               
fixup af42c27 Schema changes moved to branch.sql                                                              
pick 14045c6 Clean up code formatting                                                                         
fixup 005573e Clean up code formatting                                                                        
pick d9980f5 Remove extraneous argument                                                                       
fixup fe2c70f Remove extraneous argument                                                                      
pick 1b3a373 Refined and added tests                                                                          
pick 98f0f3e Code formatting and cleanup                                                                      
                                                                                                              
# Rebase f887fdd..98f0f3e onto f887fdd
```

whit537 commented:
The problem I'm running into is that the `park_payment_instructions` ends up in the final diff. The rebased `Refined and added tests` commit doesn't remove that method, for some reason. That makes me wonder what *else* is awry.

whit537 commented:
Alright, new strategy:

```
pick 97ead4e Preliminary commit - charging in arrears                                                         
pick 5b58b70 Schema changes moved to branch.sql                                                               
pick 14045c6 Clean up code formatting                                                                         
pick d9980f5 Remove extraneous argument                                                                       
pick 1b3a373 Refined and added tests                                                                          
pick 98f0f3e Code formatting and cleanup                                                                      
                                                                                                              
# Rebase f887fdd..98f0f3e onto f887fdd
```

I compared the diffs for the duplicate commits, and decided that the second of each ought simply to be skipped:

 - Preliminary commit - charging in arrears: 816776f9dd7db8761f7caf15001e5dd5af1138ef makes three additions that are already included with 97ead4ea8cfab5fa2db7eb157c4d0e7b2c59be8c.
 - Schema changes moved to branch.sql: the one deletion in af42c27c7801362c2fce33ec841e801a89b97250 is included in 5b58b70fbe9e40abdabdd672f1d1dd91bdedc791.
 - Clean up code formatting: 005573edea63535729aaa17f126a6c3dffab12da is the same as 14045c61eadafb72cebef807dde3cd86f3216384 ... except! for a tell-tale second `park_payment_instructions` at the end of the diff.
 - Remove extraneous argument: fe2c70f01f2b7ffd7f796576bcdb9a856edf1ba4 is the same as d9980f52a51b6eaa99c05005c5b22ed7e6733a75.

rorepo commented:
Strange. I had this method come up as a conflict both when rebasing as well as when I tried to push to the branch (and was told to do a pull first). But thought I did finally get rid of it.

whit537 commented:
Ah, nuts! The 5c23e5b489d7efc861dad2a7b93fbc07483bbdff commit *still* doesn't remove `park_payment_instructions`. :rage:

whit537 commented:
That's the only difference between the diffs.

whit537 commented:
https://github.com/gratipay/gratipay.com/commit/5c23e5b489d7efc861dad2a7b93fbc07483bbdff.diff
https://github.com/gratipay/gratipay.com/commit/1b3a373.diff

whit537 commented:
The "Code formatting and cleanup" diff is good.

https://github.com/gratipay/gratipay.com/commit/aad163e7875c6143c4b6c939d30c50e8db578860.diff
https://github.com/gratipay/gratipay.com/commit/98f0f3e.diff

whit537 commented:
How does the diff look to you, @rorepo? Are we back on track here?

rorepo commented:
Diff looked ok, but just in case, took the precaution of comparing the full file (payday.py). Matches my local copy. And no park_payment_instructions in evidence :-)

whit537 commented:
> Matches my local copy. And no park_payment_instructions in evidence :-)

Awesome. :-)

Onward with review ...

rorepo commented:
Did the same with payday.sql. Afraid it's got this
```DROP TABLE IF EXISTS participants_payments_uncharged;
CREATE TABLE participants_payments_uncharged AS
    SELECT id, giving_due
      FROM payday_payment_instructions
     WHERE 1 = 2;```
which shouldn't be there (this table was meant for ```park_payment_instructions```)

whit537 commented:
@rorepo Okay. Can I let you go ahead and remove that?

rorepo commented:
Remove that - on github itself, you mean?

whit537 commented:
@rorepo Yeah, I just meant make another commit here on this PR to sync up payday.sql with what it should be.

rorepo commented:
@whit537 Done :-) Doubt if it would've been so easy if I'd tried pushing my copy again

whit537 commented:
Cool, thanks. :)

rorepo commented:
Well, you did most of the hard work, @whit537 :-) I'm not comfortable enough with git yet, to handle this kind of a mess.

whit537 commented:
I still have no idea how we got into that mess, though. :-(

Okay!

I still want *some* kind of record of the actual deferred payment. It could be a log message during payday, or an entry in the events table, or a record in a new table. We need *something* to help with debugging when the day comes that we need to debug something around this.

Do we have adequate test coverage for the behavior of `giving_due` once a payment is satisfied?

Also, let's use `payment_due` or just simply `due` instead of `giving_due`, as the latter is with reference to the ~user's point of view, and we should use the viewpoint-agnostic term this far down in the app.

rorepo commented:
The ```giving_due``` column in the active record in ```payment_instructions``` would always have the value of the deferred payment, but I agree it would be useful to have this information readily available in something like an event log for ready reference rather than having to query the table. 
Could I have an explanation for the events table? Perhaps a sample of the records?

Thinking of test coverage, it occurs to me that there are a few more that might be useful:

1.  ```giving_due``` value is reset to 0 after a successful payment, as you suggested
2.  Carry-over of ```giving_due``` value is taken into account when checking against the charge threshold i.e. ```amount + giving_due``` is indeed used in this
3.  Things work ok when the subscription value is changed, both upwards and downwards. Here, I have a question about how the change should be handled. I'm assuming that we do not change things in mid-week. Say, if the subscription is changed on Monday from $3 to $5, we keep any existing ```giving_due```, accrued at the rate of $3/week, unchanged and only apply the $5 to subsequent weeks.

I could change ```giving_due``` to ```due``` in the meantime

I have one other question - should I be uploading yml files if they've changed?

whit537 commented:
> The `giving_due` column in the active record in `payment_instructions` would always have the value of the deferred payment[.]

This picks up with the conversation from https://github.com/gratipay/gratipay.com/issues/3378#issuecomment-126612705:

> If we have multiple records in [payment_instructions] corresponding to each individual [payment instruction] (modification), seems to me we could use the giving_due column in [payment_instructions] itself and still have a capability to trace giving_due back to individual instances of the [payment instruction].

But we change the record in `payment_instructions` when the amount of the payment changes, not simply when the payment is fulfilled (i.e., actually charged).

1. payment instruction 0: $2.00 from ~alice to the Bakelite Team
1. Payday N: $2.00 due from ~alice to Bakelite, no charge
1. payment instruction 1: $3.00 from ~alice to Bakelite
1. payment instruction 2: $2.50 from ~alice to Bakelite
1. Payday N+1: $4.50 due from ~alice to Bakelite, no charge
1. Payday N+2: $7.00 due from ~alice to Bakelite, no charge
1. payment instruction 3: $1.00 from ~alice to Bakelite
1. Payday N+3: $8.00 due from ~alice to Bakelite, no charge
1. payment instruction 4: $4.00 from ~alice to Bakelite
1. Payday N+4: $12.00 due from ~alice to Bakelite, charge!

I'm looking for an easy way to produce this history:

1. Payday N: $2.00 due from ~alice to Bakelite because of payment instruction 0
1. Payday N+1: $4.50 due from ~alice to Bakelite because of payment instruction 2
1. Payday N+2: $7.00 due from ~alice to Bakelite because of payment instruction 2
1. Payday N+3: $8.00 due from ~alice to Bakelite because of payment instruction 3
1. Payday N+4: $12.00 due from ~alice to Bakelite because of payment instruction 4
1. Payday N+4: $0.00 due from ~alice to Bakelite because of charge X

Given the existing implementation on this PR, how would we produce that history?

whit537 commented:
> Could I have an explanation for the events table? Perhaps a sample of the records?

Grep for the `add_event` function for examples of usage.

whit537 commented:
> I have one other question - should I be uploading yml files if they've changed?

Yes, but in a separate commit from the test that uses the new yaml fixture, to avoid cluttering diffs.

whit537 commented:
> Here, I have a question about how the change should be handled. I'm assuming that we do not change things in mid-week. Say, if the subscription is changed on Monday from $3 to $5, we keep any existing giving_due, accrued at the rate of $3/week, unchanged and only apply the $5 to subsequent weeks.

Correct. We modify `due` exactly once per week: during payday. Each payday has a starting timestamp, and that's [the cutoff for which payment instruction we use](https://github.com/gratipay/gratipay.com/blob/1882/sql/payday.sql#L61) during that particular payday. A payment instruction can and sometimes will change many times during the week, but whatever a payment instruction is when payday starts, that's what we'll go with.

rorepo commented:
Whatever's this conflict? Tried looking at the diff but didn't get it.

whit537 commented:
@rorepo The conflicts are due to #3698 landing on master. I'm rebasing now ...

whit537 commented:
Tentative rebase in https://github.com/gratipay/gratipay.com/compare/cia. Let's compare those commits to the ones on this PR. Does that look right? If so we can force-push that over `charge-in-arrears` and we'll be back on track here.

whit537 commented:
Alright! Rebased. :-)

Back on track, @rorepo?

rorepo commented:
Yes, looks ok, @whit537

rorepo commented:
@whit537, I implemented a new test, covering the scenario you were talking about - besides the history aspect, verifying these different permutations was necessary anyway. 

I'm afraid I had to make a design change. Kept running into an error like this:
> E       OperationalError: tuple to be updated was already modified by an operation triggered 
by the current command
E       HINT:  Consider using an AFTER trigger instead of a BEFORE trigger to propagate 
changes to other rows.

This appeared to be caused by the updates of **payday_payment_instructions** records in function ```process_payment_instruction``` in payday.sql, conflicting with the BEFORE trigger on this table. This error had come up earlier and it seemed to go away by itself (in other words, I don't know how), but it was persistent this time. I therefore introduced a new table **payment_instructions_due** for the 'pay' and 'park' updates of 'due'.

About the history bit - I'm afraid I couldn't make use of ```add_event``` itself, since payday.py does not actually have visibility of subscription level operations. Therefore I inserted records into **events** directly, from the 'pay' and 'park' points in payday.sql. I used *payday* as the event type.

At the moment, facing a problem with git push. I started with a git fetch origin and fresh checkout of origin/charge-in-arrears and made my changes on top of that. Checking manually on github, git diff, etc. suggested things were fine. Made my (2) commits, after which git status showed my local branch ahead of origin/charge-in-arrears by 2 commits. At this point, I did a rebase origin/master preparatory to pushing the changes and got the message 
> Your branch and 'origin/charge-in-arrears' have diverged,
and have 31 and 11 different commits each, respectively.

Will need some help understanding/fixing this before I push.

whit537 commented:
> At the moment, facing a problem with git push.

Hmmm ... my first hunch is that maybe you've got a local `charge-in-arrears` branch that is conflicting with `origin/charge-in-arrears`?

rorepo commented:
Well, this is how I tried to ensure things were in sync:

1. I did have a local ```charge-in-arrears branch```, in which I'd been doing most of my trials
2. When I had a reasonable idea of the implementation, I took a look at github and it had some changes in the meantime, such as the introduction of 'TheEnterprise', 'picard', etc. Therefore, I renamed (moved) my existing local branch as ```charge-in-arrears-old```, did a git fetch origin and then 'git checkout charge-in-arrears' at which point I got a message like 'branch charge-in-arrears set up to track origin/charge-in-arrears. Checking the log as well as manually looking at the code, things seemed to be in sync, as expected.
3. I then (manually) transferred my changes to the new branch and made my commits. Git status at this point informed me that my branch was 2 commits ahead of ```origin/charge-in-arrears```, as expected
4. Did git rebase origin/master. It was at this point that I got the message about branches having diverged.

I tried deleting the old branch, but that didn't change anything.

rorepo commented:
Been trying different permutations. Here's the history of the last one:
```git branch -D charge-in-arrears```
```git branch -D charge-in-arrears-old```
```git fetch origin```
```git rebase origin/master```
*Current branch master is up to date.*
```git checkout charge-in-arrears```
*Branch charge-in-arrears set up to track remote branch charge-in-arrears from origin.*
```git status```
*On branch charge-in-arrears*
*Your branch is up-to-date with 'origin/charge-in-arrears'.*
```git rebase origin/charge-in-arrears```
*Current branch charge-in-arrears is up to date.*
```git diff origin/charge-in-arrears```
no differences shown
```git diff origin/master```
shows differences
```git rebase origin/master```
*First, rewinding head to replay your work on top of it...*
*Applying: Preliminary commit - charging in arrears*
*Applying: Schema changes moved to branch.sql*
*Applying: Clean up code formatting*
*Applying: Remove extraneous argument*
*Applying: Refined and added tests*
*Applying: Code formatting and cleanup*
*Applying: Kill park_payment_instructions method*
*Applying: Trim trailing whitespace*
*Applying: Removed extraneous table script*
*Applying: Renamed payment_instructions 'giving_due' to 'due'*
*Applying: Fix whitespace typo*
```git status```
*On branch charge-in-arrears*
*Your branch and 'origin/charge-in-arrears' have diverged,*
*and have 29 and 11 different commits each, respectively.*
 * (use "git pull" to merge the remote branch into yours)*

whit537 commented:
@rorepo Okay, I'm able to reproduce your result. It's normal and expected. What has happened is that `master` has moved on since the last time we rebased this PR branch `charge-in-arrears` on it. If you scroll back through git history on a clean `charge-in-arrears` checkout then you should see that it is based on the 1884 version. After the rebase dance you describe (resulting in "29 and 11 different commits each, respectively") then you should see that `charge-in-arrears` is now has the 1887 version just before this PR's 11 commits.

That means you are doing it right! You are properly replaying this branch's commits onto the latest `master`. Now all that remains is to force-push your local, rebased `charge-in-arrears` branch back up to the GitHub remote.

Here's what a regular `git push` will look like:

```
$ git push
To git@github.com:gratipay/gratipay.com.git
 ! [rejected]        charge-in-arrears -> charge-in-arrears (non-fast-forward)
error: failed to push some refs to 'git@github.com:gratipay/gratipay.com.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
$
```

You need to *force*-push since we want to alter the history of this PR to include the latest changes on `master`. Go ahead and do this after the `dance` you [described](https://github.com/gratipay/gratipay.com/pull/3675#issuecomment-134283403):

```
$ git push --force
```

Or just:

```
$ git push -f
```

Force-pushing PR branches is fine and normal. **BUT**, force-pushing `master` itself is Very Bad&trade; and you should never do it except in very special circumstances, so make sure you only ever force-push PR branches. :-) 

I'll go ahead and let you do the force-push of the rebased PR so you get a feel for it.

Once you've rebased on `master` then you should be able to push your two additional commits with no problem.

rorepo commented:
Thanks for the clarification, @whit537. There's still something I'm doubtful about, though. Here are two extracts from history (I used git log --oneline --decorate --graph --all). It appears that the 11 commits brought in by my rebase at this point were already merged into master, under different commit ids (the first time is after version 1884). Is this ok?

![branch history 1](https://cloud.githubusercontent.com/assets/7984085/9460051/23dbb060-4afa-11e5-9015-7f5643141f79.PNG)
![branch history 2](https://cloud.githubusercontent.com/assets/7984085/9460055/2d6f17e8-4afa-11e5-96bf-2a791ef4beb3.PNG)

Also, one question about the force push - do I do this right after the rebase and then only do my additional commits and push?

whit537 commented:
> Is this ok?

Hmm ... no, that doesn't sound okay. To clarify: those history snippets are from immediately after the [steps you described above](https://github.com/gratipay/gratipay.com/pull/3675#issuecomment-134283403)?

> Also, one question about the force push - do I do this right after the rebase and then only do my additional commits and push?

It actually doesn't matter. If your force-push includes the two additional commits then that is fine.

Here are some diagrams to explain what rebasing does. It replays individual commits starting at a new base commit, and *it results in new commits*. That's why you have to force-push to change the branch (remember that branch names are simply aliases for SHA1 commit hashes).

#### First

The `charge-in-arrears` branch is based off of the commit pointed to by `master`.

![screen shot 2015-08-26 at 11 43 04 am](https://cloud.githubusercontent.com/assets/134455/9498490/206c90b8-4be8-11e5-8387-546b7a2acffc.png)

#### Second

The `master` branch is advanced to include two commits that are *not* also on `charge-in-arrears`.

![screen shot 2015-08-26 at 11 43 35 am](https://cloud.githubusercontent.com/assets/134455/9498506/2b047072-4be8-11e5-84de-8523b21d0705.png)

#### Third

Rebasing `charge-in-arrears` on the new `master` (`d`) creates new commits, 

![screen shot 2015-08-26 at 11 45 21 am](https://cloud.githubusercontent.com/assets/134455/9498508/300adaca-4be8-11e5-82cd-df7a09d7092d.png)

Because, following the rebase, one's local `charge-in-arrears` no longer shares history with the remote `charge-in-arrears`, a force-push is necessary to repoint the remote `charge-in-arrears` to `y`.

If, after the local rebase but before force-pushing, you had added a `z` commit locally, then a force-push would still have been necessary to alter the history of the remote `charge-in-arrears`, and the new `charge-in-arrears` would include `z`.

whit537 commented:
Being "on `charge-in-arrears`" means being in the history of the commit pointed to by the alias `charge-in-arrears`.

rorepo commented:
I understand rebase. But haven't been able to figure out how those duplicates got in, or how to get them out - other than perhaps cherry picking (out) these while merging.

whit537 commented:
@rorepo I noticed d28f5daa7230de74fc9407d5453d943013394866 and 1a2f382f13351b61db944e897684e7f2bc74a573 while digesting your last comment—are those your two most recent commits? I.e., are we on-track on this PR as it stands?

whit537 commented:
@rorepo Okay, I think I know what's happening: up at https://github.com/gratipay/gratipay.com/pull/3675#issuecomment-133364613 I made a second branch, `cia`, as part of testing out the rebase I was doing. If I'm reading your second history extract correctly (formatted as it is, it's a bit hard to follow), we are looking at the commit on `cia` there, *not* on `master`.

Yeah, I see the same commit hashes in my own history using `--all`:

```
| * 0208bfd (upstream/pull/3706, origin/turn-communities-off) Turn communities off
|/  
| * 9762b75 (origin/cia, cia) Fix whitespace typo
| * 95732c4 Renamed payment_instructions 'giving_due' to 'due'
| * 5bce7c8 Removed extraneous table script
| * ab653dc Trim trailing whitespace
| * 64f0f57 Kill park_payment_instructions method
| * eaa5cc2 Code formatting and cleanup
| * 096d2b9 Refined and added tests
| * 21fff58 Remove extraneous argument
| * 63e2521 Clean up code formatting
| * 84d99c6 Schema changes moved to branch.sql
| * 3582630 Preliminary commit - charging in arrears
|/  
| * 38109c0 (str) Bring back a couple lines that were rebased out
```

whit537 commented:
I like `--all`, thanks for the tip. :-)

whit537 commented:
Actually, I like `--decorate` even better. :)

rorepo commented:
Yes, @whit537, the two commits you mentioned are the final ones I made and other than for the thing about the duplicates, things are on track.

I was kind of tickled to find ```--decorate``` too. Almost understood what it showed :-)

whit537 commented:
> other than for the thing about the duplicates

Do you buy my explanation at https://github.com/gratipay/gratipay.com/pull/3675#issuecomment-135132355? :-)

rorepo commented:
Buy, yes. But understood - only after I tried the graph without the ```--all``` :-) 

So now, looks like we're all set :-)

whit537 commented:
Huzzah! So let's see about this new code now ... :-)

rorepo commented:
I'd been puzzled by ```current_payment_instructions``` - didn't see it being used anywhere. But then, forgot about it. 
Now, occurs to me that using this instead would also do away with the necessity for the first update. 

However, not sure I got this:
> let's change this test to exercise that.

whit537 commented:
> However, not sure I got this:

In other words, change queries like this:

```sql
            SELECT DISTINCT ON (participant, team) due
              FROM payment_instructions ppi
             WHERE ppi.participant = 'obama'
               AND ppi.team = 'TheEnterprise'
          ORDER BY participant, team, mtime DESC
```

into queries like this:

```sql
            SELECT due
              FROM current_payment_instructions cpi
             WHERE ppi.participant = 'obama'
               AND ppi.team = 'TheEnterprise'
```

Maybe even moving it into a method (untested):

```python
class Participant:
    [...]
    
    def get_due(self, team):
        return self.db.one("""
            SELECT due
              FROM current_payment_instructions cpi
             WHERE ppi.participant = %s
               AND ppi.team = %s
        """, (self.username, team.slug))
```

```

rorepo commented:
Ok, got it :-)

whit537 commented:
> (untested)

By which I meant, "I have not yet tested or run the following code." :)

whit537 commented:
Today is payday! Maybe we can run with this?

whit537 commented on gratipay/billing/payday.py:
```diff
@@ -354,6 +361,22 @@ def update_balances(cursor):
+              FROM payment_instructions_due pi2
+             WHERE pi.id = pi2.id
+        """)
+        # Reset older due values to zero
```
I think this is extraneous. If we use the `current_payment_instructions` view in order to get the `due` amount, then we'll always get the current `due` amount. We'll never use values from old `payment_instructions` records, so it doesn't matter whether they're zero or some other value. Right?

whit537 commented on sql/payday.sql:
```diff
@@ -55,7 +55,7 @@ CREATE TABLE payday_payments_done AS
 CREATE TABLE payday_payment_instructions AS
-    SELECT participant, team, amount
+    SELECT s.id, participant, team, amount, due
       FROM ( SELECT DISTINCT ON (participant, team) *
```
This subquery can be replaced with a query against `current_payment_instructions`.

rorepo replied:
The current subquery has this:
```WHERE mtime < %(ts_start)s```
while ```current_payment_instructions``` does not.
Can this be a problem?

whit537 replied:
Yeah, we'll need to move that `WHERE` clause up into the parent query.

rorepo replied:
As I understand it, ```ts_start``` is defined at the start of payday. This would make it 
difficult to incorporate this into the ```current_payment_instructions``` view as a condition that would be meaningful at other times as well as at payday.

I'd also like to understand the situation where ```mtime``` in ```payment_instructions``` 
could/would be >= payday ```ts_start```. e.g. can a user specify a start time that is in the future,
for a subscription? Was unable to find any such provision in the dev environment.

rorepo replied:
I've made all changes except for this particular one, because I'm not sure about the filtering by ```ts_start```. As I see it, ```payment_instructions``` can only have a record with ```mtime``` > ```ts_start``` if a new subscription is created after the payday run has been started. And if this situation comes up, we do have a problem - the ```current_payment_instructions``` view will pick up this latest record, while ```payday_payment_instructions``` (as it is now) will not. If we do update the subquery, this could have different results depending on the exact point in payday when the new subscription is created.

Once again, I've got a 'diverged' situation after rebasing with master:
>Your branch and 'origin/charge-in-arrears' have diverged,
and have 196 and 14 different commits each, respectively.

However, this seems ok after checking the log - will just have to do a forced push like last time.

whit537 replied:
> [T]he `current_payment_instructions` view will pick up this latest record[.]

Good point. Withdrawn. :-)

whit537 commented:
I've added a `get_due` method.

I saw a single test failure on [this line](https://github.com/gratipay/gratipay.com/blob/1cfe67452ce240e1272863a2da043fb4a3ff6c7f/tests/py/test_billing_payday.py#L499) under that last commit (the actual was 0), but I haven't been able to reproduce it.

whit537 commented on sql/payday.sql:
```diff
@@ -76,12 +76,27 @@ CREATE INDEX ON payday_payment_instructions (participant);
 
+UPDATE payday_payment_instructions ppi
+   SET due = s.due
+  FROM (SELECT participant, team, SUM(due) AS due
```
Why are we summing `due` here? My understanding was that `due` in `current_payment_instructions` would always be cumulative.

rorepo replied:
Well, I'm thinking of a situation like this:

1. Payday 0 - subscription amount $3, uncharged, resulting in due $3
2. Subscription amount changed to $4 i.e new record entered with amount = $4 (and due = 0)
3. Payday 1 - ```current_payment_instructions``` would pick up the latest record, with due = 0 and the existing due of $3 would be missed

In this situation, we'd need some way to transfer the existing due to the new record at step 2. Thought of using an AFTER INSERT trigger, but updating ```payday_payment_instructions``` seemed simpler.

whit537 replied:
Gotcha. I think we should update the `due` column every time a new `payment_instruction` is `INSERT`ed. We want to be able to surface that data in the UI, so we should keep it current, we shouldn't only ever update it during payday.

> Thought of using an AFTER INSERT trigger[.]

How about carrying the `due` column forward during `Participant.set_payment_instruction`?

whit537 commented on tests/py/test_billing_payday.py:
```diff
@@ -29,8 +29,91 @@ def test_payday_moves_money(self):
+        assert self.obama.get_due('TheEnterprise') == D('2.00')
+
+        self.obama.set_payment_instruction(Enterprise, '3.00')  # < MINIMUM_CHARGE
+        self.obama.set_payment_instruction(Enterprise, '2.50')  # cumulatively still < MINIMUM_CHARGE
```
Cumulative with what? Not cumulative with the 3.00 just above, since this replaces that, right?

rorepo replied:
Correct. Just the 2.50 at this point.

whit537 commented:
> Once again, I've got a 'diverged' situation after rebasing with master:

Yeah, that's normal. Everytime you rebase you're rewriting history, and you're going to have to force-push to get the remote (GitHub) to accept your rewrite.

whit537 commented:
@rorepo I rebased on master to resolve conflicts between this PR and .https://github.com/gratipay/gratipay.com/commit/aef017ab20b38fc092d2014422425aa293d764e7

whit537 commented:
That means you'll need to rebase your most recent local commits before you can push.

whit537 commented:
Alright, @rorepo, I think we just need to get your latest commits on here and then we're ready for final review! :-)

rorepo commented:
Working on it, @whit537. There were a few conflicts to be sorted out. 

Ran into a new problem in testing: 
> 'column ncc_failing of relation paydays does not exist'. 

I see schema.sql now has a 'DROP COLUMN' for this and other columns. Payday.py still refers this column - possibly a merge issue when I rebased?. Tests run ok if I comment out the drop line for this column.

whit537 commented:
@rorepo Yeah, sorry for the curveball. :-/

`ncc_failing` is one of the columns we dropped from `paydays`, so yes it's probably a rebase issue. The only reference I see to `ncc_failing` on `master` is a test that is marked `xfail`.

rorepo commented:
I see ```ncc_failing``` is directly referenced only in an ```xfail``` marked test. However, it is this bit of code in payday.py that is causing the problem for me:
>        # Record the number of failures
        cursor.one("""
            UPDATE paydays
               SET ncc_failing = %s
             WHERE ts_end='1970-01-01T00:00:00+00'::timestamptz
         RETURNING id
        """, (n_failures,), default=NoPayday)

This is in ```create_card_holds```. Could this be something that crept back in when I rebased?

whit537 commented:
> Could this be something that crept back in when I rebased?

Yeah, that got taken out in https://github.com/gratipay/gratipay.com/commit/aef017ab20b38fc092d2014422425aa293d764e7 on #3730. Try your rebase again?

whit537 commented:
I saw the same conflict when I rebased this branch on master.

rorepo commented:
Rebasing didn't make any difference. 

I uncommented the 'DROP COLUMN' before pushing - so tests will fail for the moment.

whit537 commented:
Did I introduce that code into this branch when I did my latest rebase onto master?

whit537 commented:
Blech. Looks like it. :-( It's in the diff here.

whit537 commented:
Sorry, @rorepo. My git-fu needs leveling up. :)

rorepo commented:
@whit537, it's beyond my git-fu at the moment, which is why I decided to push as is :-) 

I could try correcting on github if you could point me in the right direction.

whit537 commented:
@rorepo I'll re-rebase now that we have your latest commit on here ...

rorepo commented:
I tried commenting out that code in payday.py in my local branch - tests ran ok. So looks like that (and one more line - see last bit of messages below) is indeed the only conflict. 
However, saw another problem. Got this when I ran the tests:
> gratipay/models/participant.py:1175: redefinition of unused 'update_giving' from line 916
gratipay/models/participant.py:1207: redefinition of unused 'update_due' from line 946
gratipay/models/participant.py:1244: redefinition of unused 'update_is_free_rider' from line 999
gratipay/models/participant.py:1260: redefinition of unused 'profile_url' from line 815
gratipay/models/participant.py:1267: redefinition of unused 'get_teams' from line 822
gratipay/models/participant.py:1278: redefinition of unused 'get_old_teams' from line 833
gratipay/models/participant.py:1294: redefinition of unused 'insert_into_communities' from line 849
gratipay/models/participant.py:1312: redefinition of unused 'change_username' from line 867
gratipay/billing/payday.py:240: local variable 'n_failures' is assigned to but never used

I checked the first few methods - they are defined twice now.

whit537 commented:
Okay, I effed up the rebase. I will try to clean it up here once I get payday wrapped (https://github.com/gratipay/inside.gratipay.com/issues/329). Sorry. :-(

rorepo commented:
Good luck :-)

whit537 commented:
Thanks! :-)

whit537 commented:
@rorepo I looks like we have duplicate commits again ...

whit537 commented:
I ran this locally to get your latest commit without the two extra commits I was seeing on feeec33 (2493f711fbed2f758211858851935a8f6d6f0741 and 2493f711fbed2f758211858851935a8f6d6f0741):

```
$ git fetch
[...]
$ git cherry-pick feeec33
$ git push --force
```

whit537 commented:
And now we're down to a pyflakes error! :-)

whit537 commented:
Aaaaaaaand that's the error you were pointing out at https://github.com/gratipay/gratipay.com/pull/3675#issuecomment-137605106. :-(

whit537 commented:
It's your latest commit, a8a6bc4bc8c564d2bb100b22d0078ae1bde21e83, that adds the duplicate methods to `Participant`.

whit537 commented:
I've force-pushed a8a6bc4 off the branch, and now we're green again. What was the intention with that commit? The code doesn't seem to match what it says on the tin:

> **Updated to use current_payment_instructions view**
> current_payment_instructions view used in place of the new payment_instructions_due table which has now been removed

whit537 commented:
feeec33079972f53715ba12d8d474eabc523bfd3 and a8a6bc4bc8c564d2bb100b22d0078ae1bde21e83 are identical.

whit537 commented:
Once we land this, should we issue refunds for the remaining balances of people that we charged in advance? It would make #3539 easier.

whit537 commented:
@rorepo I think we need you to redo your last commit.

rorepo commented:
I think where my merges go wrong is when things go to conflict resolution. Either I'm not reading/handling the conflicts right or git is deficient in picking up/highlighting things like duplicated code. But my understanding of git is improving, so hopefully this won't be a perennial problem.

A few thoughts on refunding remaining balances charged in advance:

1. We'd need a threshold for this as well, otherwise we'd end up trying to refund even below the min payout value, say.
2. There's also the aspect that with small balances which would be consumed in a couple of weeks, there may not be much point in doing the refund. So maybe we can link the threshold to this, say, 2 weeks into subscription amount.
3. In any case, I think we should try to eliminate the situation where we charge (in arrears) and end up refunding excess balance right away.
4. Remembering @rohitpaulk's comment a while back, that excess balances as a result of being charged in advance were a relatively low percentage of transaction volumes - what would be the percentage of such balances once we've switched to charging in arrears?

whit537 commented:
Good work, @rorepo. :-)

!m @rorepo

rorepo commented:
:-) Glad there were no more problems.

What about the refund bit? Will it make sense to start looking at this?

whit537 commented:
@rorepo Well, we'll see if there any more problems on Thursday! :-)

I've reticketed the refund as #3763.

!m @rorepo
